### PR TITLE
Issue #143 - Default permission on install and save

### LINKED ISSUE #143: Enable default permissions for new paragraph types
As a site builder
I want by default anonymous to have view access to paragraphs
So that I don't need to add that permission each time

### PR DESCRIPTION
**Pull request by yorkshire-pudding**
Fixes #143

### CODE REVIEW

**Comment by laryn** [comment]
@yorkshire-pudding Does this PR overwrite permissions for existing Paragraphs types, or just provide defaults for new Paragraphs types? If the former that may be problematic in the case that someone has configured their Paragraphs type differently than these defaults, and then the permissions get overwritten by a default setting. (I haven't tested yet but I'm wondering about the changes in the install file.)

File: modules/paragraphs_bundle_permissions/paragraphs_bundle_permissions.install [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  // each paragraph type.
+  // Get role names.
+  $roles = user_roles();
+  // Compile the message.
+  $t = get_t();
+  $message = $t('Default permissions have been added to the %anonymous role (View) and %administrator role (All) for all existing paragraph types. Review and make changes if needed on the <a href="@url">permissions</a> page.',
+    array(
+      '%anonymous' => $roles['anonymous'],
+      '%administrator' => $roles['administrator'],
+      '@url' => paragraphs_bundle_permissions_url(),
```
Comment (laryn) [inline modules/paragraphs_bundle_permissions/paragraphs_bundle_permissions.install]: Can we use the basic permissions page URL and add a query on this link to take advantage of the new permissions filter? 
e.g. build this URL: `admin/config/people/permissions?search=paragraphs`

Then we might not need the special function at all. This link/portion of the message should only show if the user has permission to update permissions.

File: paragraphs.admin.inc [hunk 2]
```diff
@@ -244,6 +244,66 @@ function paragraphs_admin_bundle_form_submit(array $form, array &$form_state) {
...
+        '%anonymous' => $roles['anonymous'],
+        '%admin' => $roles['administrator'],
+        '%paragraph_type' => $bundle->name,
+      )
+    );
+    // Define part of the message depending on whether the user has permission
+    // to administer permissions or not.
+    if (user_access("administer permissions")) {
+      // Generate a url to the relevant section of the permissions screen.
+      $url_base = 'admin/config/people/permissions';
```
Comment (laryn) [inline paragraphs.admin.inc]: This section should be simplified/reworked to use the core permissions filter as well.

**Review by laryn — CHANGES_REQUESTED** [review]
@yorkshire-pudding I think it's looking great. In the meantime however we have a new core permissions filter that should simplify a lot of this code. I think we can strip out a lot and just build this URL when needed (e.g. when the user has permission to modify permissions):

```
admin/config/people/permissions?search=paragraphs%20type%20permissions
```

File: paragraphs.admin.inc (hunk 2 above)
Comment (laryn) [reply paragraphs.admin.inc]: @yorkshire-pudding If we do end up keeping `paragraphs_bundle_permissions_url` can we use it here as well?

**Comment by laryn** [comment]
@yorkshire-pudding Two thoughts I'd love to hear your perspective on:

1) I think we need to use something like `$admin_role = config_get('system.core', 'user_admin_role');` rather than assuming `administrator` as the role.

2) We're using the same logic structures and strings multiple times for the conditional messages, etc. Is there any way to simplify that in a reasonable way to reduce redundancy?

**Comment by yorkshire-pudding** [comment]
> @yorkshire-pudding Two thoughts I'd love to hear your perspective on:
> 
> 1. I think we need to use something like `$admin_role = config_get('system.core', 'user_admin_role');` rather than assuming `administrator` as the role.
> 2. We're using the same logic structures and strings multiple times for the conditional messages, etc. Is there any way to simplify that in a reasonable way to reduce redundancy?

Thanks for these suggestions @laryn - I hadn't considered the first one but of course, people can choose which role gets the new admin permissions.
Regarding the second, yes, that component does make sense to put into a shared function.  Two points to note:
- I've used `get_t()` so it will work with `hook_install`
- I've kept the separate `paragraphs_bundle_permissions_url()` function as this is called without a message from the Operations menu on the Paragraphs Overview page.

**Comment by laryn** [comment]
Thanks for your persistence and patience @yorkshire-pudding! Merged.

**Pull request merged**